[webkitcorepy.string_utils] Add a function that rounds seconds and minutes.

bls1999 opened the pull request:
#### 6617ffedc1ede4e9ff869ff483ec876475b2d071
<pre>
[webkitcorepy.string_utils] Add a function that rounds seconds and minutes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288808">https://bugs.webkit.org/show_bug.cgi?id=288808</a>
<a href="https://rdar.apple.com/145818734">rdar://145818734</a>

Reviewed by NOBODY (OOPS!).

Adds more customizability to the elapsed function in string_utils.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
    (elapsed): Adds options for precision (rounding digits) and shorthand.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/string_utils_unittest.py:
    (StringUtils):
    -&gt; (test_elapsed): Add more test cases.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6617ffedc1ede4e9ff869ff483ec876475b2d071

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2044 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97726 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43242 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70993 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28430 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95733 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9501 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83997 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51324 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/92226 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9193 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1630 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42570 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79516 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1603 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99748 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19783 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79998 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/92312 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20034 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79893 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79300 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23832 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1142 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12797 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19767 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19454 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22914 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21195 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
```diff
@@ -92,23 +92,31 @@ def out_of(number, base):
-        return 'less than a second'
-
-    seconds = int(round(seconds))
+def elapsed(seconds: float, precision: int = None, shorthand: bool = False):
```
Can we write a few unit tests for your changes? They should be really easy to write.

gsnedders replied:
Can we make the new arguments keyword only arguments? i.e.:

```suggestion
def elapsed(seconds: float, *, precision: int = None, shorthand: bool = False):
```

Also many of the test cases violate the type specified for `seconds`; we probably want either `Union[int, float]` or `numbers.Real`.

bls1999 committed:
[webkitcorepy.string_utils] Add a function that rounds seconds and minutes.
https://bugs.webkit.org/show_bug.cgi?id=288808
rdar://145818734

Reviewed by NOBODY (OOPS!).

Adds more customizability to the elapsed function in string_utils.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
    (elapsed): Adds options for precision (rounding digits) and shorthand.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/string_utils_unittest.py:
    (StringUtils):
    -> (test_elapsed): Add more test cases.

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
```diff
@@ -92,23 +92,31 @@ def out_of(number, base):
-    seconds = int(round(seconds))
+def elapsed(seconds: float, precision: int = None, shorthand: bool = False):
+    rounded_seconds = round(seconds, ndigits=precision)
+    if seconds < 60:
```
Given the output here is `rounded_seconds`, shouldn't this be `rounded_seconds < 60`?

Testing `elapsed(59.9)` and other values around boundaries would probably be good (likewise `60 * 60 - 0.1` and `60 * 60 * 24 - 0.1`).

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
```diff
@@ -92,23 +92,31 @@ def out_of(number, base):
+                return 'less than a second'
+        return pluralize(rounded_seconds, 'second')
+
+    seconds = int(seconds)
```
This seems potentially problematic, e.g. in the `59.9` case — `seconds` will now be `59`, while `rounded_seconds` is 60.

It's also unclear why we need this?

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
```diff
@@ -92,23 +92,31 @@ def out_of(number, base):
     hours = minutes // 60
     days = hours // 24
 
+    precise_seconds = (seconds % 60) + round(rounded_seconds - seconds, ndigits=precision)
```
What is this trying to calculate? The number of seconds?

Because in the `59.9` case currently, this will be equal to `60`, when presumably we're expecting a value in the range [0,60) (i.e., `0.0 ≤ precise_seconds < 60.0`).

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
```diff
@@ -92,23 +92,31 @@ def out_of(number, base):
     if force:
-        return result[1:] + ' and ' + pluralize(seconds % 60, 'second')
-    return pluralize(seconds % 60, 'second')
+        return result + f'{precise_seconds}s' if shorthand else f"{result[1:]} and {pluralize(precise_seconds, 'second')}"
```
Why is this `result[1:]` rather than `result`? Especially in the shorthand case, this seems wrong.

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/string_utils.py:
```diff
@@ -92,23 +92,31 @@ def out_of(number, base):
     force = False
     result = ''
-    for value, description in [(days, 'day'), (hours % 24, 'hour'), (minutes % 60, 'minute')]:
+    for value, description, letter in [(days, 'day', 'd'), (hours % 24, 'hour', 'h'), (minutes % 60, 'minute', 'm')]:
```
We should probably just be calling `divmod` above, rather than doing the division and modulo separately.
